crates/sui-core/src/accumulators/mod.rs

```diff
@@ -0,0 +1,248 @@
+            MergedValueIntermediate::SumU128U128(v1, v2) => MergedValue::SumU128U128(v1, v2),
+        }
+    }
+}
+
+/// MergedValueIntermediate is an intermediate / in-memory representation of the for
+/// accumulators. It is used to store the merged result of all accumulator writes in a single
+/// checkpoint.
+///
+/// This pattern is not necessary for fully commutative operations, since those could use MergedValue directly.
+///
+/// However, this supports the commutative-merge + non-commutative-update pattern, which will be used by event
+/// streams. In this pattern, everything within a checkpoint is merged commutativley, and then a single
+/// non-commutative update is applied to the accumulator at the end of the checkpoint.
+#[derive(Debug, Copy, Clone)]
+enum MergedValueIntermediate {
+    SumU128(u128),
+    SumU128U128(u128, u128),
+}
+
+impl MergedValueIntermediate {
+    // Create a zero value with the appropriate type for the accumulator value.
+    fn zero(value: &AccumulatorValue) -> Self {
+        match value {
+            AccumulatorValue::Integer(_) => Self::SumU128(0),
+            AccumulatorValue::IntegerTuple(_, _) => Self::SumU128U128(0, 0),
+        }
+    }
+
+    fn accumulate_into(&mut self, value: AccumulatorValue) {
+        match (self, value) {
+            (Self::SumU128(v1), AccumulatorValue::Integer(v2)) => *v1 += v2 as u128,
+            (Self::SumU128U128(v1, v2), AccumulatorValue::IntegerTuple(w1, w2)) => {
+                *v1 += w1 as u128;
+                *v2 += w2 as u128;
+            }
+            _ => {
+                fatal!("invalid merge");
+            }
+        }
+    }
+}
+
+// TODO(address-balances): This currently only creates a single accumulator update transaction.
+// To support multiple accumulator update transactions, we need to:
+// - have each transaction take the accumulator root as a "non-exclusive mutable" input
+// - each transaction writes out a set of fields that are disjoint from the others.
+// - a barrier transaction must be added to advance the version of the accumulator root object.
+//   The barrier transaction doesn't do any field writes. This is necessary in order to provide
+//   a consistent view of the system accumulator state. When the version of the accumulator
+//   root object is advanced, we know that all accumulator state updates prior to that version
+//   have been applied.
+pub fn create_accumulator_update_transactions(
+    epoch_store: &AuthorityPerEpochStore,
+    checkpoint_height: u64,
+    cache: Option<&dyn TransactionCacheRead>,
+    ckpt_effects: &[TransactionEffects],
+) -> Vec<TransactionKind> {
+    let epoch = epoch_store.epoch();
+    let Some(accumulator_root_obj_initial_shared_version) = epoch_store
+        .epoch_start_config()
+        .accumulator_root_obj_initial_shared_version()
+    else {
+        debug!("accumulator root object does not exist, skipping accumulator update");
```

Comment:
Since we only call in here if it's enabled, we could unwrap here. Or we could even pass the init shared version from the caller